Require executor name when calling scheduleWithFixedDelay

jaymode opened the pull request:
The ThreadPool#scheduleWithFixedDelay method does not make it clear that all scheduled runnable instances will be run on the scheduler thread. This becomes problematic if the actions being performed include blocking operations since there is a single thread and tasks may not get executed due to a blocking task.

This change includes a few different aspects around trying to prevent this situation. The first is that
the scheduleWithFixedDelay method now requires the name of the executor that should be used to execute the runnable. All existing calls were updated to use Names.SAME to preserve the existing behavior.

The second aspect is the removal of using ScheduledThreadPoolExecutor#scheduleWithFixedDelay in favor of a custom runnable, ReschedulingRunnable. This runnable encapsulates the logic to deal with rescheduling a runnable with a fixed delay and mimics the behavior of executing using a ScheduledThreadPoolExecutor and provides a ScheduledFuture implementation that also mimics that of the type returned by a ScheduledThreadPoolExecutor.

Finally, an assertion was added to BaseFuture to detect blocking calls that are being made on the scheduler thread.

bleskes commented on core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -364,10 +370,13 @@ public ScheduledExecutorService scheduler() {
      * @param command the action to take
      * @param interval the delay interval
-     * @return a ScheduledFuture who's get will return when the task is complete and throw an exception if it is canceled
+     * @return a ScheduledFuture who's get will block until the task has been canceled and then once canceled will throw a
```
can we add a word about the implications of using SAME? i.e. this will run on the schedule thread, of which there is only one and should only do super light weight things.

bleskes commented on core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -971,4 +980,147 @@ public void close() throws IOException {
         return threadContext;
     }
+
+    static class ReschedulingRunnable implements Runnable {
```
can we use AbstractRunnable ? this will give us handy onAfter and onFailure methods.

bleskes commented on core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -971,4 +980,147 @@ public void close() throws IOException {
+            }
+        }
+
+        class CancellableScheduledFuture implements ScheduledFuture<Void> {
```
I really appreciate the effort to mimic the original future returned by scheduleWithFixedDelay. However, since we never really use the get() and rely on the cancel, I suggest we implement a super simple AtomicBoolean based cancel and throw UnsupportedOperationException for all the rest. These things are super complex (compared, delay, exception capturing). Let's wait until we need them and keep things simple till then. WDYT?

jaymode replied:
I'm ok with simplifying if we do not return a ScheduledFuture as it feels wrong/misleading to return a class where we implement a single method and throw an exception for the other methods. 

Instead, we can return our own class that can be used to cancel and that is it. We could call it `CancellableTask` or something and just implement a `cancel` method. What do you think about that?

bleskes replied:
Works for me. Maybe just have an interface called Cancellable  - it seems CancellableTask is taken.

bleskes commented:
Thx @jaymode . I left some comment and a suggestion

jaymode commented:
@bleskes I pushed a new commit to address your feedback

dakrone commented on core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -1005,4 +1011,82 @@ public ThreadContext getThreadContext() {
+
+    public static boolean assertNotScheduleThread() {
+        assert Thread.currentThread().getName().contains("scheduler") == false :
+            "Expected current thread [" + Thread.currentThread() + "] to not be the scheduler thread. Reason: " + "[Blocking operation]";
```
Don't really need the last `+` in this?

jaymode replied:
nope, leftover from IDE wrapping lines automatically

bleskes replied:
strictly speaking, we can log the reason here as we don't know why people called us. We need to either pass a parameter (reason) or remove the last part.

dakrone commented:
LGTM, left one suuuuper minor comment

bleskes commented:
@jaymode what's the state of this?

jasontedor commented:
> @jaymode what's the state of this?

I'm taking this PR over.

jaymode commented:
@bleskes I merged master to bring this back up to date. I am waiting for a second review, maybe you can take a look to ensure I addressed all of your comments?

bleskes commented on core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -1011,4 +1017,82 @@ public ThreadContext getThreadContext() {
+        boolean isCancelled();
+    }
+
+    static class ReschedulingRunnable extends AbstractRunnable {
```
can we add java docs to this and the main methods?

bleskes commented on core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -1011,4 +1017,82 @@ public ThreadContext getThreadContext() {
+
+        private final Runnable runnable;
+        private final TimeValue interval;
+        private final String name;
```
can this executor? name is what we often give the tasks..

bleskes commented on core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -1011,4 +1017,82 @@ public ThreadContext getThreadContext() {
+        public void onFailure(Throwable t) {
+            // if a runnable throws, the default behavior of ScheduledThreadPoolExecutor is to stop execution
+            run = false;
+            threadPool.logger.warn("failed to run scheduled task [{}] on thread pool [{}]", runnable.toString(), name);
```
can we log the exception?

bleskes commented on core/src/test/java/org/elasticsearch/monitor/jvm/JvmGcMonitorServiceSettingsTests.java:
```diff
@@ -150,4 +153,19 @@ private static void execute(Settings settings, BiFunction<Runnable, TimeValue, S
         }
     }
 
+    interface TriFunction<S, T, U, R> {
```
😄

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+
+        ThreadPool threadPool = new ThreadPool("counting-runnable");
+        try {
+            Cancellable cancellable = threadPool.scheduleWithFixedDelay(countingRunnable, TimeValue.timeValueMillis(100L), Names.GENERIC);
```
can we make this more evil and throw exceptions here and there?

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+            assertTrue(cancellable.isCancelled());
+
+            // we cancel but there could be a currently running iteration, so we buffer by one value...
+            final boolean stillRunning = awaitBusy(() -> counter.get() > currentCount + 1, 250L, TimeUnit.MILLISECONDS);
```
yeah, this is tricky. I instead I think we can run a test that cancels from the running task and makes sure it is never run again? we can then decide whether to cancel immediately or after a few iterations

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+        final AtomicInteger counter = new AtomicInteger(0);
+        final Runnable countingRunnable = counter::incrementAndGet;
+
+        ThreadPool threadPool = new ThreadPool("cancelling-runnable");
```
maybe set up the thread pool in before / after?

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+        // verify that there were no more interactions with the threadpool such as scheduling again
+        verifyNoMoreInteractions(threadPool);
+
+        assertTrue(cancellable.isCancelled());
```
I'm confused - where did we cancel?

bleskes replied:
cancel that :) I figured it out.

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+        startLatch.await();
+
+        // this call was made during scheduleFirstRun
+        verify(threadPool, times(1)).schedule(delay, Names.GENERIC, reschedulingRunnable);
```
I'm confused by this test - it's checking that a if two runs are blocked by a comon latch, they don't complete and reach reschudeling? also why don't we have two rescheduling when the thread and the scheduled task finish?

jaymode replied:
this test is really a validation of the behavior of the runnable in that it will not reschedule itself until after completion of the run method. I'll add some comments to make this a bit more clear

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+
+        Cancellable cancellable = reschedulingRunnable.scheduleFirstRun();
+        if (randomBoolean()) {
+            reschedulingRunnable.run();
```
why do we do this?

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+        }
+        cancellable.cancel();
+
+        IllegalStateException exception = expectThrows(IllegalStateException.class, reschedulingRunnable::scheduleFirstRun);
```
schedule first run should be called twice, regardless of cancellation right? can we check for that using an assertion? we don't really have to test it as there is no way for the user to achieve this when the class is not exposed.

bleskes commented:
@jaymode the implementation looks great. The tests need some tightening up as discussed. Also, I know it was previous behavior, but I think make scheduleWithFixedDelay reschedule on error will make it simpler to use, do what we expect of it (we currently don't deal with errors right) and allows us to use it in more places and make things like refresh scheduling code simpler. Can we move to that model? Again, thanks a lot for picking this up.

jaymode commented:
@bleskes I made the changes we discussed. If an exception occurs, the task will still be rescheduled and tried to make the tests more evil

bleskes commented on core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -696,4 +701,102 @@ public ThreadContext getThreadContext() {
+     * {@link ScheduledThreadPoolExecutor#scheduleWithFixedDelay(Runnable, long, long, TimeUnit)} semantics as an exception there would
+     * terminate the rescheduling of the runnable.
+     */
+    static class ReschedulingRunnable extends AbstractRunnable implements Cancellable {
```
can we mark this as final, just in case?

bleskes commented on core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -696,4 +701,102 @@ public ThreadContext getThreadContext() {
+
+        @Override
+        public void onFailure(Throwable t) {
+            threadPool.logger.warn("failed to run scheduled task [{}] on thread pool [{}]", t, runnable.toString(), executor);
```
nit: runnable.toString() is implied no? no need to make it explicit is it?

bleskes commented on core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -696,4 +701,102 @@ public ThreadContext getThreadContext() {
+
+        @Override
+        public void onRejection(Throwable t) {
+            if (threadPool.logger.isDebugEnabled()) {
```
I think we should cancel here, no?  currently, the onAfter will be called resulting in an endless loop.

also, if we're shutting down, this might result in needless warnings, so maybe check if we're canceled and if so only log  a debug log instead of warn?

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+
+    @After
+    public void shutdown() {
+        threadPool.shutdownNow();
```
be nice and use terminate?

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+
+            if (doneLatch.getCount() == 0) {
+                runAfterDone.set(true);
+                fail("this runnable was cancelled by itself so it should never run again!");
```
hmm. does this actually fail the test? we capture throwables in ReschedulingRunnable and just log them?

bleskes replied:
oh, I see the assert at the end. Maybe just warn log instead?

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+
+            final int count = counter.decrementAndGet();
+
+            if (shouldThrow && randomBoolean()) {
```
what does this adds us in terms of testing coverage? isn't it just adding complexity?

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+
+        // we rely on concurrency to check for the case where a run after cancel happens, but we can rarely do some waiting
+        if (rarely()) {
+            assertFalse(awaitBusy(runAfterDone::get, 2L, TimeUnit.SECONDS));
```
I don't follow this one? we just checked it's false?

jaymode replied:
this is hard to test since the runnable could run after we check it being false. This rarely awaits to make sure it isn't run. If you don't think it is necessary, I can remove it.

bleskes replied:
oh,  I see now. So I would adapt the comment. maybe "rarely wait and make sure the task is not run at the next interval".

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+        }
+    }
+
+    public void testConcurrentCancel() throws Exception {
```
what do we test here exactly?

jaymode replied:
this tests multiple threads calling cancel(). The threads all wait for each other to have started and the test to signal them to run and each of them calls cancel.

bleskes replied:
I get that, but do we test that multiple thread calling cancel doesn't result in an exception? since we just set a boolean to false on cancel, this feels like an overkill, no?

jaymode replied:
yeah so I will remove the test since we moved to the simplified version

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+
+        final Runnable runnable = () -> {
+            try {
+                if (randomBoolean()) {
```
can we fix this into a boolean before we start? will be easier to reproduce / know what happened.

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+        };
+
+        final Cancellable cancellable = threadPool.scheduleWithFixedDelay(runnable, TimeValue.timeValueMillis(20L), Names.GENERIC);
+        boolean failed = awaitBusy(() -> throwableRef.get() != null, 200, TimeUnit.MILLISECONDS);
```
I'm not a fan of this. we effectively always wait. Can we just rely on the `assertNull(throwableRef.get())` at the end of the test? it may have some false positives but if we get something wrong, we'll know soon enough

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+
+        final Object o = new Object();
+        future.futureDone(o);
+        boolean done = awaitBusy(() -> objectRef.get() != null, 200, TimeUnit.MILLISECONDS);
```
if we use double futures (one that you have already and another to pass back the results) we won't have to busy wait here.

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -0,0 +1,287 @@
+        assertTrue(cancellable.isCancelled());
+    }
+
+    public void testCancellingRunnable() throws Exception {
```
I'm fine we keeping the test of canceling from the runnable task, but can we also have a test for external canceling where post cancel call the runnable is a runned at most once? Note that to do this you will again be in that situation where you wait for something that shouldn't happen, causing the test to be slow - I'm fine with a direct check that will sometime cause false positives.

jaymode replied:
Do you want more than what `testThatRunnableIsRescheduled` does? Maybe we just add the runAfterDone boolean there?

bleskes commented:
@jaymode looks great. Thx! I left some suggestions to improve testing and one semantic change (cancel on rejection)

jaymode commented:
@bleskes this is ready for review again. I addressed all of your previous feedback

bleskes commented on core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java:
```diff
@@ -221,67 +219,90 @@ public void run() {
-        assertFalse(failed);
-        assertNull(objectRef.get());
-        assertNull(throwableRef.get());
+        expectThrows(TimeoutException.class, () -> resultsFuture.get(0L, TimeUnit.MILLISECONDS));
```
just check resultsFuture.isDone()?

bleskes commented:
LGTM! left one nit. Thanks so much Jay!

jaymode commented:
This was back-ported to the 2.4 branch in https://github.com/elastic/elasticsearch/commit/ebd900e955f0bb09bf2cec24a1816ed7728307fc

nithril commented:
My Spring scheduler thread name contains `scheduler` hence I get an assertion error from elastic. The comparison seems to me a bit too broad.  Maybe the elastic thread pool prefix can be more specific.

bleskes commented:
@nithril can you elaborate more about your setup? are you using the transport client in your spring application?

nithril commented:
Elastic 2.4.1
Java 1.8

I'm using a `NodeClient` (used for the integration testing). Actually, the exception occurs in my test after I renamed my scheduler. A scheduled task is executed which performs a blocking search.

The stack is the following: 
```
Caused by: java.lang.AssertionError: Expected current thread [Thread[website-scheduler_Worker-2,5,main]] to not be the scheduler thread. Reason: [Blocking operation]
	at org.elasticsearch.threadpool.ThreadPool.assertNotScheduleThread(ThreadPool.java:1089)
	at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:119)
	at java_util_concurrent_Future$get.call(Unknown Source)
```
